### DISCUSSION TITLE
Correct nuxt hook for generating a full-static payload server-side before building the client?

### DISCUSSION THREAD

**Discussion by MuhammadM1998** [nuxt/nuxt#16051] [Questions]
After suffering a lot I got playwright to run on `nuxt dev` and `nuxt generate` wothout errors.
Now I don't know which hook is the right one to achieve the same result in `generate` as in `dev`

I'm using Playwright as a [nitro plugin ](https://nuxt.com/docs/guide/directory-structure/plugins) to scrape data from a website then use it in the frontend.

`server/plugins/playwright`
```javascript
import { chromium } from 'playwright-core';
import bundledChromium from 'chrome-aws-lambda';
import { db } from '../db';

export default defineNitroPlugin(async (nitroApp) => {
  try {
    // Setup
    const browser = await Promise.resolve(bundledChromium.executablePath).then(
      (executablePath) => {
        if (!executablePath) {
          // local execution
          return chromium.launch({ headless: false });
        }
        return chromium.launch({
          executablePath,
          headless: true,
          args: [
            '--no-sandbox',
            '--disable-setuid-sandbox',
            '--disable-dev-shm-usage',
            '--single-process',
          ],
        });
      }
    );
    // Scraping
    const url = 'https://google.com/';
    const page = await browser.newPage();
    await page.goto(url);

    await page.close();
    await browser.close();

    // Returning data
    console.log('Works');
    db.push('Works!');
  } catch (error) {
    console.log(error);
    db.push(error);
  }
});
```

`server/routes/playwright.js`
```javascript
import { db } from '../db';
export default defineEventHandler(() => db);
```

`server/db.js`
```javascript
export const db = [];
```

`pages/index.vue`
```vue
<script setup>
  const { data } = await useFetch('/playwright');
</script>

<template>
  <div>
    <p>Nuxt with Playwright Demo</p>
    <p>{{ data }}</p>
  </div>
</template>
```

### Behaviour
- Running `nuxt dev` works fine, playwright works correctly and `works` is pushed to `db` then it's rendered on index.js on `localhost`
- Running `nuxt generate` works without erros, but running `nuxt preview` and opening `localhost` an empty array is rendered instead of `['Works']` (as in dev)

### Extra Context
I've tried the [nitro hooks](https://nuxt.com/docs/api/advanced/hooks#nitro-hooks-runtime-server-side) but it didn't work correctly, and on `nuxt generate` this is logged but still on visiting `localhost:3000` after `nuxt preview` I get an empty array rendered
```
i Client built in 2037ms
i Building server...  
√ Server built in 755ms
√ Generated public .output/public
i Initializing prerenderer
i Prerendering 3 initial routes with crawler
├─ / (3244ms)
├─ /200.html (3141ms)
├─ /404.html (2819ms) 
├─ /_payload.js (3111ms)
```

**Comment by danielroe** [comment]
Is this still the case if you run `npx serve .output/public` instead of `nuxt preview`?

**Reply by MuhammadM1998** [reply]
I found a temp solution, I attached it in the end of the reply.

`npx serve .output/public` does the same, I think the problem is not in `nuxt preview` rather than in the build itself as I checked `.output/public/payload.js` and It has an empty array instaed of `['Works']`.

This is what's logged on running `nuxt preview`
Isn't `Working dir: .output` and `Starting preview command: npx serve -s ./public`
the same as `npx serve .output/public` ?
```
yarn run v1.22.19
$ nuxt preview
Nuxi 3.0.0
i Node.js version: 14.21.1
i Preset: nitro-prerender
i Working dir: .output
i Loading .env. This will not be loaded when running the server in production.
i Starting preview command: npx serve -s ./public

INFO: Accepting connections at http://localhost:3000.
```

**I found out that running the exact code as a route handler works!**
Removing `plugins/playwright.js` and adjusting  `routes/playwright.js` workis in `dev` and `generate`
```javascript
import { chromium } from 'playwright-core';
import bundledChromium from 'chrome-aws-lambda';
import { db } from '../db';

export default defineEventHandler(async () => {
  const browser = await Promise.resolve(bundledChromium.executablePath).then(
    (executablePath) =>
      !executablePath
        ? chromium.launch({ headless: false }) // Local execution
        : chromium.launch({ executablePath }) // Remote execution
  );

  const context = await browser.newContext();
  const page = await context.newPage();
  await page.goto('https://www.google.com/');
  await page.waitForTimeout(2000);
  await browser.close();

  db.push('works');
  return db;
});
```

It's fine for me to use since I'll build it on generate only it but it's annoying in dev as it runs on every refresh
